proposal for disabled Quick Starts (#715)
openshift/enhancements: enhancements/console/quick-starts.md

* Networking
* ...

> For Quick Starts we need a mechanism for their creation and publishment. We also need a mechanism to allow cluster administrators to control the visibility of any of the Quick Starts in console.

Reviewer: ```suggestion
For Quick Starts we need a mechanism for their creation and publishment. We also need a mechanism to allow cluster administrators to control the visibility of Quick Starts in the console UI.
```

2. Make users understand what are the steps needed to achieve their goals.
3. Provide new CRD format for writing the Quick Starts.
4. Have a repository for the out-of-the-box Quick Starts, those that describe how to install an operator or go though standard workflow.

> 5. Allow cluster admins to control the visibility of any of the Quick Starts to users on the cluster.

Reviewer: ```suggestion
5. Allow cluster administrators to hide specific Quick Starts from displaying in the UI for all users.
```

Author: done

### Non-Goals

> 1. Introducing a UI to list out disabled Quick Starts without a yaml editor.

Reviewer 2: ```suggestion
1. Introducing a UI to list out disabled Quick Starts without a YAML editor.
```

Author: done

## Proposal

* Introduce CRD format that will be used for writing Quick Starts.
* Introduce default repository for Quick Starts.

> * Extend the existing `operator.openshift.io~v1~Console` CRD. It has a `spec.customization` property where the list of Quick Start Ids to be disabled can be added.

Reviewer 2: ```suggestion
* Extend the existing `operator.openshift.io~v1~Console` CRD. It has a `spec.customization` property where the list of Quick Start names to be disabled can be added.
```

Author: done

As a operator creator I want to provide operator consumers with a guide on how to install and user the my operator.

#### Story 5

> As an administrator of an Openshift cluster, I should be able to hide any of the Quick Starts so that it is not available to anyone on the cluster.

Reviewer: ```suggestion
As an administrator of an OpenShift cluster, I want to hide specific Quick Starts from displaying in the UI for all users.
```

Author: done

    console.openshift.io/name: explore-serverless
```

#### Disabling Quick Starts

`operator.openshift.io/v1` / `Console` CRD provides a `spec.customization` property where an admin will list the Quick Starts to be disabled.

> Console operator CRD with disabled Quick Starts:

Reviewer 2: I assume this will disable quick starts with this `metadata.name` or `console.openshift.io/name` annotation (for translated quick starts). If so, I'd call that out specifically here.

Reviewer: Good catch! Yes this will match the `console.openshift.io/name` annotation. It should only match against `metadata.name` if there is no annotation.

Author: done

    console.openshift.io/name: explore-serverless
```

#### Disabling Quick Starts

> 1. `operator.openshift.io~v1~Console` CRD provides a `spec.customization` property where an admin will add a property `disabledQuickStarts` with the list of Quick Start Ids to be hidden.
> 2. Extend the openshift-console operator so that the console resource data are available in the console ConfigMap and provide them as SERVER_FLAGS.
> 3. The Quick Starts listed in `window.SERVER_FLAGS.disabledQuickStarts` will not be visible on the UI

Reviewer: Remove implementation details that are only relevant to console consuming this data.

```suggestion
`operator.openshift.io/v1` / `Console` CRD provides a `spec.customization` property where an admin will list the Quick Starts to be disabled.
```

Author: done

Console operator CRD with disabled Quick Starts:
```yaml
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
  ...
spec:
  customization:
    brand: online

>     disabledQuickStarts:
>       - 'serverless-application'

Reviewer: I'm leaning towards creating a specific section for quick starts in case we have some future need.
```suggestion
    quickStarts:
      disabled:
        - 'serverless-application'
```

Author: I agree. It makes sense to have a specific section accommodate future quick start related requirements.
Thanks :)
updated the YAML accordingly.
